include/envoy/grpc/async_client.h

```diff
@@ -169,7 +169,12 @@ template <class ResponseType> class TypedAsyncStreamCallbacks : public AsyncStre
   virtual void onReceiveMessage(std::unique_ptr<ResponseType>&& message) PURE;
 
   void onReceiveMessageUntyped(ProtobufTypes::MessagePtr&& message) override {
-    onReceiveMessage(std::unique_ptr<ResponseType>(dynamic_cast<ResponseType*>(message.release())));
```

Comment:
I think the fix here should be slightly different. It seems reasonable to check if the message is the wrong type (I'm actually surprised this hasn't come up before and/or we aren't already doing this), but I think you should try the dynamic cast first, and if it's null, then go into the error path, and consider this an error response. Also you will need some tests for this case.